tee/src/lib.rs

```diff
@@ -0,0 +1,594 @@
+		/// Enclave already assigned to a cluster
+		EnclaveAlreadyAssigned,
+		/// Enclave not assigned to a cluster
+		EnclaveNotAssigned,
+		/// Enclave does not exists
+		EnclaveDoesNotExists,
+		/// Cannot assign to same cluster
+		CannotAssignToSameCluster,
+		/// Internal logical error
+		InternalLogicalError,
+		/// Assigning an operator to an invalid enclaveId
+		AssigningOperatorForUnknownEnclaveId,
+		/// Unknown enclave operator account
+		UnknownEnclaveOperatorAccount,
+		/// Invalid IAS sign certificate
+		InvalidIASSigningCert,
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+
+		#[pallet::weight(T::WeightInfo::register_enclave())]
+		pub fn register_enclave(
+			origin: OriginFor<T>,
+			enclave_address: Vec<u8>,
+			api_uri: Vec<u8>,
+		) -> DispatchResultWithPostInfo {
+			let account = ensure_signed(origin)?;
+
+			ensure!(api_uri.len() < T::MaxUriLen::get().into(), Error::<T>::UriTooLong);
+			ensure!(api_uri.len() > T::MinUriLen::get().into(), Error::<T>::UriTooShort);
+
+			let has_registered_index = EnclaveIndex::<T>::get(account.clone());
+
+			// Verify if the origin does not already have an enclave address associated with it.
+			match has_registered_index {
+				Some(enc_id) => {
+					let enc = EnclaveRegistry::<T>::get(enc_id).unwrap();
+					ensure!(
+						enc.enclave_address != enclave_address,
+						Error::<T>::EnclaveAddressAlreadyRegisteredtoTheAccount
+					)
+				}
+				_ => {}
+			}
+
+			let (enclave_id, new_id) = Self::new_enclave_id()?;
+			// Needs to have enough money
+			let imbalance = T::Currency::withdraw(
+				&account,
+				T::EnclaveFee::get(),
+				WithdrawReasons::FEE,
+				KeepAlive,
+			)?;
+			T::FeesCollector::on_unbalanced(imbalance);
+
+			let enclave = Enclave::new(api_uri.clone(), enclave_address);
+
+			EnclaveIndex::<T>::insert(account.clone(), enclave_id);
+			EnclaveRegistry::<T>::insert(enclave_id, enclave);
+			EnclaveIdGenerator::<T>::put(new_id);
+
+			let reg_enclaves:  BoundedVec<EnclaveId, T::MaxRegisteredEnclaves> = <RegisteredEnclaves<T>>::get();
+			let reg_enclaves = reg_enclaves
+				.try_mutate(|v| v.push(7))
+				.unwrap();
+			<RegisteredEnclaves<T>>::put(reg_enclaves);
+
+			Self::deposit_event(Event::AddedEnclave { account, api_uri, enclave_id });
+			Ok(().into())
+		}
+
+		/// Removes an enclave from the system
+		/// Origin- operator account address
+		/// If the enclave is assigned, it will be placed in queue for tech committee approval
+		/// If enclave is not already assigned, he can exit without permission.
+		#[pallet::weight(T::WeightInfo::unregister_enclave())]
+		pub fn unregister_enclave(
+			origin: OriginFor<T>,
+		) -> DispatchResultWithPostInfo {
+			let account = ensure_signed_or_root(origin)?;
+			match account {
+				Some(account_id) => {
+					let enclave_id  = EnclaveIndex::<T>::get(account_id.clone()).unwrap();
+
+					let un_reg_enclaves:  BoundedVec<EnclaveId, T::MaxUnRegisteredEnclaves> = <UnRegisteredEnclaves<T>>::get();
```

Comment:
don't worry about using full names, it's more readable and it does not change the final code -> unregistered_enclaves for example